backend/src/test/java/com/google/minicrm/data/CampaignTest.java

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.minicrm.data;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.users.User;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Provides Unit tests for the Campaign data object.
+ */
+@RunWith(JUnit4.class)
+public final class CampaignTest {
+
+  private static final String TEST_USER_ID = "testUserId";
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void newCampaign_fromValidCampaignAsEntity_producesEquivalentValidCampaign()
+      throws Exception {
+    User testUser = new User("email", "authDomain", TEST_USER_ID);
+    Key advertiserKey = Advertiser.generateKey(testUser);
+    Campaign campaign = new Campaign(advertiserKey, 1, "campaign1");
+
+    Campaign convertedCampaign = new Campaign(campaign.asEntity());
+
+    assertEquals(campaign, convertedCampaign);
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```

Comment:
Didn't know about this for throwing exception. Alternatively, I think you can use "assertThrows".

Reply:
I think I am doing it this way because I'm using JUnit4 and assertThrows is only in JUnit 5. Should I upgrade to JUnit5?

Reply:
I think it is available in JUnit4 as well. I think it needs to be version 4.13 or higher, and then you can use org.junit.Assert.assertThrows: https://stackoverflow.com/questions/156503/how-do-you-assert-that-a-certain-exception-is-thrown-in-junit-4-tests

Reply:
Gotcha, was using JUnit 4.12 haha, will update.